urLink Landing Page

anxiubin opened the pull request:
## TypeScript 적용
- 공통 컴포넌트 및 함수 에 타입 정의

## i18next 사용
- 다국어 지원을 위해 i18next 라이브러리 사용
- 현재 영어, 한국어 버전 지원

## react-awesome-reveal 사용
-  Intersection Observer API를 사용한 라이브러리를 이용해 스크롤 애니메이션 적용

## 반응형 웹 지원
- 미디어쿼리 사용
- 375px, 576px, 768px 1024px 기준으로 반응형 적용

## 스타일
- antd, styled-components, clsx 사용

## 코드리뷰 적용
- 폴더 구조 수정
```
src
└ baseLayout
└ components
└ pages
└ typings
└ utils
└ locales
└ router
```
- 전역 타입 선언
```ts
declare type Language = "ko" | "en"
```
- i18next에서 사용하는 t 의 type import
```ts
import { withTranslation, WithTranslation } from 'react-i18next';
const HistorySection: React.FC<WithTranslation> = ({ t: translate }) => {}
```

anxiubin committed:
✨ CRA로 프로젝트 세팅 및 router 세팅

anxiubin committed:
✨ TypeScript config 세팅

anxiubin committed:
✨ react-i18next 사용해서 다국어 처리

anxiubin committed:
💄 글로벌 스타일 추가 및 필요한 이미지 & 폰트 업로드

anxiubin committed:
✨ common 컴포넌트들 생성

anxiubin committed:
✨ Header, Footer, 공통으로 사용되는 MiddleBlock 컴포넌트 생서

anxiubin committed:
✨ Home 페이지에 들어가는 섹션 컴포넌트들 생성

anxiubin committed:
💄 반응형 디자인 적용

anxiubin committed:
✨ 버튼 클릭시 크롬 웹스토어 링크 열기

anxiubin committed:
✨ 특정 위치에서 Header 컴포넌트의 position을 fixed 로 변경

anxiubin committed:
♻️ ScrollToTop 버튼 스타일 변경

anxiubin committed:
♻️ Header padding 값 변경

anxiubin committed:
♻️ 모바일 버전에서 각 섹션마다 텍스트는 위로, 이미지는 아래로 고정해달라는 요청 반영

anxiubin committed:
💄 글로벌 스타일 수정

anxiubin committed:
♻️ 추가된 이미지 적용

anxiubin committed:
✨ Header에 다국어 버튼 추가

anxiubin committed:
♻️ 모바일 버전 이미지 변경 및 사소한 디자인들 수정

anxiubin committed:
♻️ 오타 수정

anxiubin committed:
✨ 번역된 영어 텍스트들 다국어에 적용

anxiubin committed:
♻️ 리드미 업데이트

leesoo7595 commented on src/common/Button/styles.ts:
```diff
@@ -0,0 +1,24 @@
+import styled from "styled-components"
+
+export const StyledButton = styled("button")<any>`
```
any -> 이부분을 props로 들어가는 프로퍼티가 들은 객체로 지정시켜주거나, interface를 하나 위에 만들어서 지정해주면 어떨까요???

yjkwon07 replied:
이건 모든 styles.ts에 적용 사항이 되겠네요.

anxiubin replied:
넵 any로 정의한 부분 모두 수정해놓겠습니다!

yjkwon07 commented on src/common/types.ts:
```diff
@@ -0,0 +1,30 @@
+export interface ContainerProps {
```
- types.ts를 만들지 말고 해당 컴포넌트에서 interface를 만드는것이 유지보수성이 더 잘 된다고 판단 됩니다. 
- 또한 네이밍 할때도 IProps로 정의할 수 있어 네이밍 생각을 줄여 줄 수 있습니다.

anxiubin replied:
넵 아래와 같이 전체 컴포넌트 모두 컴포넌트 내에서 Interface 로 타입 지정하고 IProps로 네이밍 통일하겠습니다.

```js
interface IProps {
	padding?: string
}

const Container: React.FC<IProps> = ({ padding, children }) => (
	<StyledContainer padding={padding}>{children}</StyledContainer>
)
```

yjkwon07 commented on src/common/types.ts:
```diff
@@ -0,0 +1,30 @@
+export interface ContainerProps {
+	padding?: string
+	children: React.ReactNode
```
- children이 필요하다면, 해당 컴포넌트에 FC<IProps>로 정의하게 되면, 자동으로 children이 오버라이팅이 됩니다.

anxiubin replied:
넵 전체 컴포넌트 모두 수정하겠습니다

yjkwon07 commented on src/common/types.ts:
```diff
@@ -0,0 +1,30 @@
+	name?: string
+	imgSrc?: string
+	children: React.ReactNode
+	onClick?: () => void
```
- ButtonProps에서는  event가 넘오온다는 props정의가 필요할거 같아요.
- (e: any | React.MouseEvent<HTMLButton...>) => void

anxiubin replied:
아래와 같이 수정해놨습니다
```js
onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void
```

yjkwon07 commented on src/common/utils/getWindow.ts:
```diff
@@ -0,0 +1,26 @@
+export function isWindow(obj: any) {
```
- 해당 utils는 common과 같은 폴더 depth 위치를 옮기는게 어떤가요? utils라는 자체가 자가 lib를 만드는 개념이라 같은 위치에서 공통으로 사용한다는 의미를 주는게 좋을거 같아요.
- 파일이름은 getWindow인데, getScroll도 포함되어 있네여, 관심 분리로, 파일을 따로 관리하여 의미론적 해석 유추를 할 수 있는게 좋아 보입니다.
- 모든건, index.ts에서 export로 내려주는 방식 (recommend)

anxiubin replied:
폴더 구조 변경했으며, 파일 분리했습니다.
또한 utils에 각각 폴더 생성후 index.ts를 만들어 관리하는 방식으로 변경했습니다.

yjkwon07 commented on src/common/utils/getWindow.ts:
```diff
@@ -0,0 +1,26 @@
+	return obj !== null && obj !== undefined && obj === obj.window
+}
+
+export const getScroll = (
```
- getScroll 사용하는 장점을 알 수 있을까요?
- 해당 로직을 2군데 사용하긴하는데, 해당 로직에서 위치값을 받아오는게 더 이상적이라고 생각했습니다.
- 만약, 모든 컴포넌트에서 적용하게끔 만든거라면, 모든 컴포넌트 coverage를 할 수 있는지 궁금합니다.

anxiubin replied:
utils로 따로 빼서 여러 차례 사용시 코드를 줄일 수 있다는 장점이 있고,
antd에 있는 코드 참고해서 가져왔는데, 아마 모든 컴포넌트 커버 가능할 것으로 예상됩니다.
[참고 링크](https://github.com/ant-design/ant-design/blob/13a829d20c8a1a83c3bfa1f9d7c30d95659f5b57/components/_util/getScroll.tsx)

yjkwon07 commented on public/index.html:
```diff
@@ -0,0 +1,28 @@
+		<meta property="og:url" content="urLink" />
+		<meta property="twitter:title" content="urLink" />
+		<meta property="twitter:description" content="urLink" />
+		<meta property="twitter:site" content="https://urlink-official.com/" />
```
- meta 태그 검색에서 정보를 좀 더 노출 될 수 있도록 보강이 좀 필요해 보이네요.
- 이건 같이 고민해 봐요.

anxiubin replied:
넵 이 부분은 카톡방에서 말씀드렸다시피 보완해야하고, 건우님도 같이 고민해보신다고 했습니다

yjkwon07 commented on src/common/ScrollToTop/index.tsx:
```diff
@@ -0,0 +1,41 @@
+		return () => {
+			window.removeEventListener("scroll", checkScrollTop)
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
- checkScrollTop 넣어도 상관 없어 보이긴 하네여.
- 한번만 적용해서 다시 add를 안하기 위함이면 패스 할게요.
- eslint 막는 주석이 많을 수록, 불필요한 코드라인 수가 늘어남이 보여지는거 같아서 없애는 방향을 생각해 봤어요.

anxiubin replied:
디펜던시에 넣으면 렌더링 될때마다 add 하게 되서 eslint 막는 주석이 필요할 것 같아요

yjkwon07 commented on src/components/Footer/index.tsx:
```diff
@@ -0,0 +1,51 @@
+} from "./styles"
+
+const Footer = ({ t }: any) => {
+	const handleChange = (language: string) => {
```
- 사용하지 않은 함수는 지워도 될거 같아요

anxiubin replied:
넵 코드 정리하겠습니다

yjkwon07 reviewed (commented):
- common이라는 폴더를 좀더 분리해서
- component
  |_ common (공통 컴포넌트)
- utils
- typings (전체 타입 declare하는 공간)

🤔 AboutSection 에서 Feature 폴더를 만들고 그 안에 index.tsx 컴포넌트를 만들어야 할까요? (urlink-extension 에서는 이러한 방식으로 사용함) 

---

- 🤔 button 클릭시 링크로 가는 로직에 대한 title 혹은 a 태그로 변환하는게 어떤지
- 🤔 image import 하여 base64값이 좋을지, 경로를 매핑하여 불러오는게 좋을지에대한 판단
- 🤔  keyframes를 컴포넌트에서 정의하는게 좋을지에대한 판단

---

- 🤔 tsconfig에서 path alias를 정의해할지 고민중 => webpack에서도 alias 설정도 해야함
- 🤔 prettier, eslint를 적용해야할까 고민중
Ex.
  - prettier => line 120 over
  - elsint => import order

yjkwon07 commented on src/components/MiddleBlock/index.tsx:
```diff
@@ -0,0 +1,46 @@
+	paddingBottom,
+	children,
+}: MiddleBlockProps) => {
+	if (animation === "reveal") {
```
- types 정의 필요. animation: "reveal" |  ""

anxiubin replied:
```ts
export interface IProps extends WithTranslation {
	animation?: "reveal" | ""
	paddingBottom?: string
}
```
위와 같이 수정했습니다

yjkwon07 commented on src/pages/Home/HistorySection/index.tsx:
```diff
@@ -0,0 +1,62 @@
+
+import { withTranslation } from "react-i18next"
+
+const moveLinkAnimation = keyframes`
```
- 위에 언급한 styles파일에서 관리하는게 어떤지에 대한 의견입니다.

anxiubin commented on src/components/MiddleBlock/index.tsx:
```diff
@@ -0,0 +1,46 @@
+
+import { MiddleBlockSection, ContentWrapper } from "./styles"
+
+const customAnimation = keyframes`
```
스타일 코드는 분리하는게 깔끔해보일 것 같긴 하네요. styles에 정의 후 import 해서 사용할게요

anxiubin committed:
♻️ 영어 번역 수정

anxiubin committed:
♻️ 컴포넌트 내에서 interface 로 타입 지정하고, IProps 로 네이밍 통일

anxiubin committed:
♻️ 스타일에 내려주는 Props 타입 정의

anxiubin commented on src/components/Header/index.tsx:
```diff
@@ -0,0 +1,88 @@
+	return (
+		<HeaderSection
+			className={clsx({
+				fixed: Boolean(fixHeader),
```
넵 지우겠습니다

anxiubin commented on src/components/Header/index.tsx:
```diff
@@ -0,0 +1,88 @@
+		>
+			<Container>
+				<Row justify="space-between">
+					<LogoContainer to="/" aria-label="homepage">
```
```js
export const LogoLink = styled(Link)`
```
위와 같이 수정했습니다

anxiubin committed:
✨ eslint, prettier, vscode settings 추가

anxiubin committed:
✨ meta tag 세팅

anxiubin commented on src/pages/Home/CardsSection/index.tsx:
```diff
@@ -0,0 +1,70 @@
+	t: any
+}
+
+const CardsSection = ({ t }: CardsSectionProps) => {
```
react-i18next 에서 제공하는 인터페이스를 사용하여 타입 문제 해결했습니다.

인터페이스를 확장할 경우 아래와 같이 사용하면 될 것으로 보이고,
```ts
import { withTranslation, WithTranslation } from 'react-i18next';

interface MyComponentProps extends WithTranslation {
  bar: 'baz';
}
```
확장할 인터페이스가 없다면 아래와 같이 사용하면 될 것 같습니다.

```ts
import { withTranslation, WithTranslation } from 'react-i18next';

const HistorySection: React.FC<WithTranslation> = ({ t }) => {}
``` 

[react-i18next 깃허브 참고](https://github.com/i18next/react-i18next/blob/master/test/typescript/withTranslation.test.tsx)

anxiubin replied:
translate 로 네이밍 변경했습니다.

```ts
const IntroSection: React.FC<WithTranslation> = ({ t: translate }) => { 
...
	<p>{translate(Contents.introSection.text[0])}</p>
}
```

anxiubin committed:
♻️ react-i18next 에서 제공하는 인터페이스를 사용하여 타입 문제 해결

- 참고 : https://github.com/urLink-DEV/urLink-homepage/pull/1#discussion_r665241724
- props 네이밍을 t => translate로 변경

anxiubin commented on src/components/Header/index.tsx:
```diff
@@ -0,0 +1,88 @@
+								show: Boolean(fixHeader),
+							})}
+						>
+							<Button type="small" onClick={onClick}>
```
`GetStartedLink` 로 컴포넌트명 변경하고 a 태그로 변경했습니다.

anxiubin committed:
✨ d.ts 파일 생성 후 전역 타입 선언

anxiubin committed:
♻️ 폴더 구조 수정

src
└ baseLayout
└ components
└ pages
└ typings
└ utils

anxiubin committed:
✨ getLocalStorageItem 함수 추가 및 적용

- 현재 선택된 언어에 따라 크롬 웹스토어 링크 변경해주기

anxiubin commented on src/components/Header/index.tsx:
```diff
@@ -0,0 +1,88 @@
+		)
+	}
+
+	const handleChange = (language: string) => {
```
`src/typings/index.d.ts` 파일 내에 아래와 같이 선언하였습니다.

```ts
declare type Language = "ko" | "en"
```

anxiubin committed:
♻️ 이미지는 소스 경로 대신 import 해서 사용하는 것으로 변경

anxiubin committed:
♻️ utils 폴더 구조 변경

anxiubin committed:
✨ Footer에 문의(contact) 버튼 추가

anxiubin committed:
♻️ GetStarted버튼 스타일 업데이트

anxiubin committed:
♻️ eslint 적용

anxiubin committed:
♻️ QA 의견 반영 : CardsSection 상단 여백 추가
